google/cloud/forseti/common/gcp_api/_base_repository.py

```diff
@@ -488,6 +488,25 @@ def execute_query(self, verb, verb_arguments):
     def _execute(self, request):
         """Run execute with retries and rate limiting.
 
+        Args:
+            request (object): The HttpRequest object to execute.
+
+        Returns:
+            dict: The response from the API.
+        """
+        try:
+            return self._execute_request(request)
+        except exceptions.RefreshError as e:
+            # If there is problem refreshing the token, we will recreate a new
+            # Credential object and use that to refresh the request.
+            LOGGER.exception(e)
+            self._credentials, _ = api_helpers.get_google_default_credentials()
```

Comment:
I'm not sure this is enough. The credential is used to create an authenticated http object in https://github.com/GoogleCloudPlatform/forseti-security/pull/1833/files#diff-ad7b8ba466c514dddccaa5f11b5a3d5aR325.

The http object is already included in the request.

The flow is:

BaseRespositoryClient -> Create credential -> Initialize GCPRepository instances for the API -> Create a thread local cached http object shared by all GCPRepository instances across all APIs (except for gsuite) -> use authenticated http object to call API methods.

Each thread has its own cached http object, which has a self.credentials object that is a pointer to the original credentials, it won't see the replaced credentials.

You could use delattr(self._local, 'http'), although that will only fix the current thread and not any other running threads.

Speaking of threads, you should not replace self._credentials without getting a lock to ensure it's not actively being used in another thread.

Reply:
Looking through https://github.com/GoogleCloudPlatform/google-auth-library-python-httplib2/blob/master/google_auth_httplib2.py#L178 and https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/master/google/auth/compute_engine/credentials.py#L35, I don't think it's the credentials that needs to be replaced, I think it's the local http object that needs to be replaced.

In fact, with google auth, I wonder if it even makes sense to do the caching any more, that was created because of oauth2client. You could try to disable the caching in self.http so it always returns a new AuthorizedHttp object.

Reply:
Actually, it looks like https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/master/google/auth/compute_engine/credentials.py#L35 is not thread safe. If multiple threads call refresh at the same time, it could corrupt the object, that may be the root cause of our issue.

Reply:
That would explain the 400 response from the metadata server.